src/pybind/mgr/volumes/fs/volume.py

```diff
@@ -865,23 +865,35 @@ def _get_src_path(self, vol_handle, dst_group, dst_subvol):
         if src_group_name == '_nogroup':
             src_group_name = None
 
-        if src_group_name != dst_group.groupname:
-            with open_subvol_in_group(self.mgr, vol_handle, self.volspec,
-                                      src_group_name, src_subvol_name,
-                                      SubvolumeOpType.CLONE_SOURCE) as src_subvol:
-                src_path = src_subvol.snapshot_data_path(src_snap_name).decode('utf-8')
+        try:
+            if src_group_name != dst_group.groupname:
+                with open_subvol_in_group(self.mgr, vol_handle, self.volspec,
+                                          src_group_name, src_subvol_name,
+                                          SubvolumeOpType.CLONE_SOURCE) as src_subvol:
+                    src_path = src_subvol.snapshot_data_path(src_snap_name).decode('utf-8')
 
-        else:
-            with open_subvol(self.mgr, vol_handle, self.volspec, dst_group,
-                             src_subvol_name, SubvolumeOpType.CLONE_SOURCE) as \
-                             src_subvol:
-                src_path = src_subvol.snapshot_data_path(src_snap_name).decode('utf-8')
+            else:
+                with open_subvol(self.mgr, vol_handle, self.volspec, dst_group,
+                                 src_subvol_name, SubvolumeOpType.CLONE_SOURCE) as \
+                                 src_subvol:
+                    src_path = src_subvol.snapshot_data_path(src_snap_name).decode('utf-8')
+        except VolumeException as e:
+            if e.errno != -errno.ENOENT:
+                raise
+
+            log.debug(f'snapshot "{src_snap_name}" which is/was being cloned to create subvolume '
```

Comment:
So, how does the cloner thread respond in such a case (when snapshot is deleted)? I think the thread pretty much keeps to retrying till the clone is canceled.

Reply:
The cloner thread enter failed state and terminates.

Reply:
Are you sure the thread itself fails? It seems like we update the clone state to "failed" and then detach the clone the clone which removed the index pointer.

Reply:
IIUC, the threads finishes too. From what I recall `start_clone_sm()` runs until `finished` and return value of `handle_clone_failed()` sets `finished`.  After this all methods in callstack of thread finish, resulting in async job coming to an end.

Reply:
(cc @kotreshhr)

I guess this change looks fine 👍

Reply:
> IIUC, the threads finishes too. From what I recall `start_clone_sm()` runs until `finished` and return value of `handle_clone_failed()` sets `finished`. After this all methods in callstack of thread finish, resulting in async job coming to an end.

That is disastrous - one failed clone takes down the thread. So, four failed clones will results in no further clones :/

Are you sure that's the case? What's the point of having a failed state then?

Reply:
I was talking in context to 1 job, 1 thread. In case of multiple jobs, thread will being next job.

Reply:
> So, how does the cloner thread respond in such a case (when snapshot is deleted)? I think the thread pretty much keeps to retrying till the clone is canceled.

I think we have checks which won't allow snapshot deletion when clone in progress. But if the snapshot is deleted from backend, then the cloner thread would end up in infinite loop trying to clone the subvolume.

Reply:
> I think we have checks which won't allow snapshot deletion when clone in progress. But if the snapshot is deleted from backend, then the cloner thread would end up in infinite loop trying to clone the subvolume.

Oh, I thought it marks the clone failed and moved forward with the next clone. So, if the thread is retying over and over again, but the stats aren't getting accounted due to this change, what side effects do we foresee?

Reply:
I tried this scenario locally. I launched 2 clone jobs after limiting number of threads to 1. Before first one could finish, I deleted the snapshot that was the source of first clone. This resulted in first clone being reported as "failed"  and second clone proceeding and reaching completion.

I'll tryit once more and post the results here.

Reply:
Clone 1 begins and Clone 2 is waiting -
```
cephfs1
└── volumes
    ├── _deleting
    ├── _index
    │   └── clone
    │       ├── bba2237b-cf69-4748-993f-afbb3b114651 -> /volumes/_nogroup/ss1c1
    │       └── e49ec666-029a-47a3-9eda-9071e7e5e2ce -> /volumes/_nogroup/ss1c2
    ├── _nogroup
    │   ├── ss1c1
    │   │   ├── 63bdc68e-302f-47af-92ae-3a8befff4e86
    │   │   │   └── file1
    │   │   └── .meta
    │   ├── ss1c2
    │   │   ├── 53b1efdc-412c-4f87-ba5c-dd97303fa249
    │   │   └── .meta
    │   ├── sv1
    │   │   ├── .meta
    │   │   └── .trash
    │   └── sv2
    │       ├── ba505e61-fdd6-4bb0-96ff-8ef5cdf11989
    │       │   ├── file1
    │       │   ├── file10
    │       │   ├── file2
    │       │   ├── file3
    │       │   ├── file4
    │       │   ├── file5
    │       │   ├── file6
    │       │   ├── file7
    │       │   ├── file8
    │       │   └── file9
    │       └── .meta
    └── _:sv2.meta
```

Just before deleting snapshot -
```
cephfs1
└── volumes
    ├── _deleting
    ├── _index
    │   └── clone
    │       ├── bba2237b-cf69-4748-993f-afbb3b114651 -> /volumes/_nogroup/ss1c1
    │       └── e49ec666-029a-47a3-9eda-9071e7e5e2ce -> /volumes/_nogroup/ss1c2
    ├── _nogroup
    │   ├── ss1c1
    │   │   ├── 63bdc68e-302f-47af-92ae-3a8befff4e86
    │   │   │   ├── file1
    │   │   │   ├── file3
    │   │   │   ├── file4
    │   │   │   └── file9
    │   │   └── .meta
    │   ├── ss1c2
    │   │   ├── 53b1efdc-412c-4f87-ba5c-dd97303fa249
    │   │   └── .meta
    │   ├── sv1
    │   │   ├── .meta
    │   │   └── .trash
    │   └── sv2
    │       ├── ba505e61-fdd6-4bb0-96ff-8ef5cdf11989
    │       │   ├── file1
    │       │   ├── file10
    │       │   ├── file2
    │       │   ├── file3
    │       │   ├── file4
    │       │   ├── file5
    │       │   ├── file6
    │       │   ├── file7
    │       │   ├── file8
    │       │   └── file9
    │       └── .meta
    └── _:sv2.meta

14 directories, 21 files
```


Clone 2 begins -
```
cephfs1
└── volumes
    ├── _deleting
    ├── _index
    │   └── clone
    │       └── e49ec666-029a-47a3-9eda-9071e7e5e2ce -> /volumes/_nogroup/ss1c2
    ├── _nogroup
    │   ├── ss1c1
    │   │   ├── 63bdc68e-302f-47af-92ae-3a8befff4e86
    │   │   │   ├── file1
    │   │   │   ├── file3
    │   │   │   ├── file4
    │   │   │   └── file9
    │   │   └── .meta
    │   ├── ss1c2
    │   │   ├── 53b1efdc-412c-4f87-ba5c-dd97303fa249
    │   │   │   ├── file1
    │   │   │   ├── file3
    │   │   │   └── file4
    │   │   └── .meta
    │   ├── sv1
    │   │   ├── .meta
    │   │   └── .trash
    │   └── sv2
    │       ├── ba505e61-fdd6-4bb0-96ff-8ef5cdf11989
    │       │   ├── file1
    │       │   ├── file10
    │       │   ├── file2
    │       │   ├── file3
    │       │   ├── file4
    │       │   ├── file5
    │       │   ├── file6
    │       │   ├── file7
    │       │   ├── file8
    │       │   └── file9
    │       └── .meta
    └── _:sv2.meta

14 directories, 23 files
```

Clone 2 finishes -
```
$ sudo tree -a cephfs1
cephfs1
└── volumes
    ├── _deleting
    ├── _index
    │   └── clone
    ├── _nogroup
    │   ├── ss1c1
    │   │   ├── 63bdc68e-302f-47af-92ae-3a8befff4e86
    │   │   │   ├── file1
    │   │   │   ├── file3
    │   │   │   ├── file4
    │   │   │   └── file9
    │   │   └── .meta
    │   ├── ss1c2
    │   │   ├── 53b1efdc-412c-4f87-ba5c-dd97303fa249
    │   │   │   ├── file1
    │   │   │   ├── file10
    │   │   │   ├── file2
    │   │   │   ├── file3
    │   │   │   ├── file4
    │   │   │   ├── file5
    │   │   │   ├── file6
    │   │   │   ├── file7
    │   │   │   ├── file8
    │   │   │   └── file9
    │   │   └── .meta
    │   ├── sv1
    │   │   ├── .meta
    │   │   └── .trash
    │   └── sv2
    │       ├── ba505e61-fdd6-4bb0-96ff-8ef5cdf11989
    │       │   ├── file1
    │       │   ├── file10
    │       │   ├── file2
    │       │   ├── file3
    │       │   ├── file4
    │       │   ├── file5
    │       │   ├── file6
    │       │   ├── file7
    │       │   ├── file8
    │       │   └── file9
    │       └── .meta
    └── _:sv2.meta

```

Note clone index was empty above -
```
$ sudo ls -la cephfs1/volumes/_index/clone
total 1
drwx------. 2 root root 0 Aug 14 13:23 .
drwx------. 3 root root 0 Aug 14 13:22 ..
```

Reply:
... and what's the `clone status` for `ss1c1`?

Reply:
it's `failed`.

```
$ ./bin/ceph fs clone status a ss1c1
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2024-08-14T13:30:33.382+0530 7f1e9b4006c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-08-14T13:30:33.391+0530 7f1e9b4006c0 -1 WARNING: all dangerous and experimental features are enabled.
{ 
  "status": {
    "state": "failed",
    "source": {
      "volume": "a",
      "subvolume": "sv1",
      "snapshot": "ss1"
    },
    "failure": {
      "errno": "2",
      "error_msg": "snapshot 'ss1' does not exist"
    }
  }
}
```

Reply:
And `complete` for `ss1c2` -
```
$ ./bin/ceph fs clone status a ss1c2
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2024-08-14T13:31:10.518+0530 7f00d08006c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-08-14T13:31:10.526+0530 7f00d08006c0 -1 WARNING: all dangerous and experimental features are enabled.
{ 
  "status": {
    "state": "complete"
  }
```